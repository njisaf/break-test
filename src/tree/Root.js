import { types } from 'mobx-state-tree';

import Lane from './Lane';
import Card from './Card';
import { type } from 'os';

//Here is our Root, the top of the Treee.
//The temptation is to start defining branches and leaves right here and noww, but we'll resist this and stay as flat as possible for as long as possiblee.
//We are worming our way into the heart of another application, and begin by replacing the Reducer.

const Root = types.model('Root', {
    loading: false,
    lanes: types.array(Lane),
    cards: type.array(Card)
})
.actions(self => {
    return {
        initializeLanes({lanes}) {
            console.log('lanes', lanes)
            self.lanes = lanes;
        },

        setLoading(boolean) {
            self.loading = boolean;
        },
    }
})
.views(self => {
    return {
        get isLoading() {
            return self.loading === true;
        },
    }
})

export default Root;