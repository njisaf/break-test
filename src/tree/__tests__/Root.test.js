import { getSnapshot, onSnapshot } from 'mobx-state-tree';
import Root from '../Root'

describe('Root', () => {

    const data = {
        lanes: [{
            id: 'lane1',
            title: 'Planned Tasks',
            label: '2/2',
            cards: [{
                    id: 'Card1',
                    title: 'Write Blog',
                    description: 'Can AI make memes',
                    label: '30 mins'
                },
                {
                    id: 'Card2',
                    title: 'Pay Rent',
                    description: 'Transfer via NEFT',
                    label: '5 mins',
                    metadata: {
                        sha: 'be312a1'
                    }
                }
            ]},
            {
                id: 'lane2',
                title: 'Completed',
                label: '0/0',
                cards: []
            }
        ]
    }

    it('can toggle .loading to true and false', () => {
        const item = Root.create();
        const states = [];
        onSnapshot(item, snapshot => {
            states.push(snapshot)
        })
        
        expect(item.loading).toBe(false);

        item.setLoading(true);
        expect(item.loading).toBe(true);
        expect(item.isLoading).toBe(true);

        expect(getSnapshot(item)).toMatchSnapshot();
        expect(states).toMatchSnapshot();
    })

    it('receives an array of lanes with attached cards, and adds both cards and lanes to the board', () => {
        const item = Root.create();
        const states = [];
        onSnapshot(item, snapshot => {
            states.push(snapshot)
        })

        expect(item.lanes.length).toBe(0);

        item.initializeBoard(data);

        expect(item.lanes.length).toBeGreaterThan(0);
    })
})