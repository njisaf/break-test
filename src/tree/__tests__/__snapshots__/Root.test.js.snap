// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root creates an snapshot of the model 1`] = `
Object {
  "loading": true,
}
`;

exports[`Root creates an snapshot of the model 2`] = `
Array [
  Object {
    "loading": true,
  },
]
`;
